### PR TITLE
Feature/ms 2325/auto targeting

### PR DESCRIPTION
**Pull request by thelightcosine**
Adds Automatic Targeting to all Exploits. 
If the Exploit does not have an already defined Automatic Target, we will try to
select a target based on what we know about the RHOST's Operating system to see if we can pick a reasonable target.

## Verification

List the steps needed to make sure this thing works

- [x] Land https://github.com/rapid7/metasploit_data_models/pull/167 first
- [x] `bundle install`
- [x] `rake spec`
- [x] **VERIFY** all specs pass
- [x] Have a Windows host saved in your MSF database
- [x] Start `msfconsole`
- [x] `use exploit/multi/http/tomcat_mgr_upload`
- [x]  `set Target 0`
- [x] `show options`
- [x] **VERIFY** it shows the target as 'Automatic'
- [x] `set RHOST <ip of your windows host>`
- [x] `show options`
- [x] **Verify** the target now says 'Windows Universal

### CODE REVIEW

File: spec/file_fixtures/modules/exploits/auto_target_linux.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+require 'msf/core'
+
+class MetasploitModule < Msf::Exploit::Remote
+  include Exploit::Remote::Tcp
+  Rank = ManualRanking
+
+  def initialize(info = {})
+    super(update_info(info,
+                      'Name'           => 'Exploit Auto-Targeting for Linux',
+                      'Description'    => %q{ This module is a test bed for automatic targeting for Windows exploits. },
```
Comment (jvoisin) [inline spec/file_fixtures/modules/exploits/auto_target_linux.rb]: `s/Windows/Linux/` ;)

**Comment by dmohanty-r7** [comment]
`spec/lib/msf/core/handler/reverse_http_spec.rb:29` fails locally for me

**Comment by busterb** [comment]
Any chance you could please rebase this and get rid of all of the merge commits please? It's hard to follow the commit history.

**Comment by dmohanty-r7** [comment]
Targeting a Windows 7 64 bit box:

```
msf > db_nmap 1.1.1.1

msf > hosts

Hosts
=====

address         mac  name                               os_name   os_flavor  os_sp   purpose  info  comments
-------         ---  ----                               -------   ---------  -----   -------  ----  --------
1.1.1.1          windowshost.cool.com   Unknown                      device
```

```
msf > use exploit/multi/http/tomcat_mgr_upload
msf exploit(tomcat_mgr_upload) > set Target 0
Target => 0
msf exploit(tomcat_mgr_upload) > show options

Module options (exploit/multi/http/tomcat_mgr_upload):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   HttpPassword                   no        The password for the specified username
   HttpUsername                   no        The username to authenticate as
   Proxies                        no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST                          yes       The target address
   RPORT         80               yes       The target port
   SSL           false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI     /manager         yes       The URI path of the manager app (/html/upload and /undeploy will be used)
   VHOST                          no        HTTP server virtual host


Exploit target:

   Id  Name
   --  ----
   0   Automatic
```

```
msf exploit(tomcat_mgr_upload) > set RHOST 1.1.1.1
RHOST => 1.1.1.1
msf exploit(tomcat_mgr_upload) > show options

Module options (exploit/multi/http/tomcat_mgr_upload):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   HttpPassword                   no        The password for the specified username
   HttpUsername                   no        The username to authenticate as
   Proxies                        no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST         1.1.1.1      yes       The target address
   RPORT         80               yes       The target port
   SSL           false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI     /manager         yes       The URI path of the manager app (/html/upload and /undeploy will be used)
   VHOST                          no        HTTP server virtual host


Exploit target:

   Id  Name
   --  ----
   0   Automatic
```

I'm not sure the windows host is getting recognized as Windows.

**Comment by dmohanty-r7** [comment]
```
msf > use exploit/multi/http/tomcat_mgr_upload
msf exploit(tomcat_mgr_upload) > set Target 0
Target => 0
msf exploit(tomcat_mgr_upload) > show options

Module options (exploit/multi/http/tomcat_mgr_upload):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   HttpPassword                   no        The password for the specified username
   HttpUsername                   no        The username to authenticate as
   Proxies                        no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST                          yes       The target address
   RPORT         80               yes       The target port
   SSL           false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI     /manager         yes       The URI path of the manager app (/html/upload and /undeploy will be used)
   VHOST                          no        HTTP server virtual host


Exploit target:

   Id  Name
   --  ----
   0   Automatic


msf exploit(tomcat_mgr_upload) > set RHOST 1.1.1.1
RHOST => 1.1.1.1
msf exploit(tomcat_mgr_upload) > show options

Module options (exploit/multi/http/tomcat_mgr_upload):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   HttpPassword                   no        The password for the specified username
   HttpUsername                   no        The username to authenticate as
   Proxies                        no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST         1.1.1.1      yes       The target address
   RPORT         80               yes       The target port
   SSL           false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI     /manager         yes       The URI path of the manager app (/html/upload and /undeploy will be used)
   VHOST                          no        HTTP server virtual host


Exploit target:

   Id  Name
   --  ----
   2   Windows Universal
```

**Comment by dmaloney-r7** [comment]
@bcook-r7 uhh the rebase made things worse...

**Comment by dmaloney-r7** [comment]
yay now everything is happy and green!

**Comment by busterb** [comment]
Happy to help. Being able to cross-push between accounts sure is fun!

**Comment by dmohanty-r7** [comment]
## Release Notes
This PR adds the ability for exploit targets (without an already defined Auto Target) to be automatically selected based on known information about the target hosts's OS in the database. The auto target is determined as soon as an rhost is defined in module configuration.

**Pull request merged**